### ISSUE TITLE
Allow instruments to play once

### ISSUE THREAD

**Issue by tmhglnd** [tmhglnd/mercury-playground#34]
Explore the options to have instruments not loop but only trigger once upon evaluation. Syntax could look like:
`new sample <name> time(off)` or `time(noLoop)`. Or maybe it should be specific keyword like `once sample <name>`

Only need to consider when the sound of the instrument will be triggered. 
1. Trigger it immediately when the code is evaluated (not synchronized with the global transport)
2. Schedule it to be triggered at the beginning of the next bar
3. Specify the triggering moment as an offset argument, for example `time(off 1/4)` would trigger the sound a quarter note later then the beginning of the bar.

**Comment by loopier** [comment]
Throwing some thoughts:
- `new sample <name> time(...) loop(times)` -- not sure if once would be `loop(0)` or `loop(1)`.

Does Mercury allow for syntax like `new sample <name> once`? Then you wouldn't have to create a new keyword. I think having just 3 keywords with `set`, `new` and `list` is a very cool feature. But I guess `once` makes sense.

**Comment by tmhglnd** [comment]
Thanks for the thoughts! `loop(0)` or `loop(1)` could be an idea also yes. Or `loop(off)`. I think in that case loop(0) would be only playing once, since it means no looping.

`new sample <name> once` is not possible with the current grammar. After the `<name>` it always expects a function in the form`func()`. I could consider the idea but this would require a rework of the grammar so potentially more work then having something like `time(off)` or `loop(0)`.

**Comment by loopier** [comment]
If `loop(0)` is once, what would `loop(1)` be, twice?

**Comment by tmhglnd** [comment]
Yeah I see how that could be confusing, where `loop(1)` would mean the sound plays twice (1 time, and then 1 repetition from the loop). 

Well it could be an idea that `loop(1)` and `loop(0)` is actually the same, so `0` and `1` just plays the sound once (so you know when you type `loop(4)` you hear the sound 4x, once + 3 repetitions. 

Then the default would be `loop(infinity)` or `loop(on)` I guess.

**Comment by loopier** [comment]
> * Will it again loop another n-amount of times? Is it then up to the coder to remove/comment the code to be sure that the next time it is evaluated the loop does not restart.
> 
Yeah, I think that's the most intuitive.

> * What will happen if a loop(32) is running, but you evaluate the code again, does it have to keep track of the amount of loops it already played and continue from there, or does it start over?
> 
I'd say it starts over. But others might find that counter-intutive :/

> * What will loop(16) mean? is that 16 hits, 16 bars? if `time(1/32)` then `loop(32)` would mean something different then with `time(1/4)`
I'd say `loop(16)` is 16 hits, or events.


> * Also there is actually already a `new loop <name>` (which is basically a sample, but the soundfile is stretched to the length of a bar so you could play for example a drumbeat synced to the tempo), so I'm not sure if this will lead to some confusion if you can have an instrument named `loop` but also all instruments have a `loop()` function.

Yes, it sounds confusing hehe

**Comment by tmhglnd** [comment]
This is now possible in the latest version in the playground. It's still "kind of" experimental but I wanna see how it works out. You can play an instrument once by adding `once(on)` or `once(true)` or `once(1)` to the line. It will then play once. The moment when it will play is determined by the `time()`. Examples:

The default time = `1/1` so if you evaluate this the harp will play once at the beginning of a bar. Try evaluating again at a random moment to hear it waits till the next bar to be triggered.
```java
set tempo 100
new sample harp_down once(on)

// just some other sound as reference
new sample hat_808 time(1/4)
```

The time is set to `2/1` meaning the harp will only play on bars multiple of 2. The first evaluation will always be the beginning. But if you randomly evaluate after that you'll have to wait till a multiple of 2 bars.
```java
set tempo 100
new sample harp_down once(on) time(2/1)

new sample hat_808 time(1/4)
```

The time is set to `2/1` but also the offset is changed to `7/8`. This means the sound will be triggered on a multiple of 2 bars and after 7/8 notes. You should hear first 4 hihat hits, than the trigger of the harp between the 4th and 1st count.

```java
set tempo 100
new sample harp_down once(on) time(2/1 7/8)

new sample hat_808 time(1/4)
```

It is possible to trigger things beyond one bar by setting the offset greater than `1/1`.
```java
set tempo 100
new sample harp_down once(on) time(2/1 7/8)
new sample metal_lo once(on) time(2/1 12/8)
new sample gong_lo once(on) time(2/1 2/1)

new sample hat_808 time(1/4)
```
